### PR TITLE
If a `content-disposition` header is set, use that rather than the URI to set the filename.

### PR DESCRIPTION
**Pull request by ndarilek**

### CODE REVIEW

File: lib/waffle/file.ex [hunk 1]
```diff
@@ -126,7 +133,28 @@ defmodule Waffle.File do
 
   defp request(remote_path, options, tries \\ 0) do
     case :hackney.get(URI.to_string(remote_path), [], "", options) do
-      {:ok, 200, _headers, client_ref} -> :hackney.body(client_ref)
+      {:ok, 200, headers, client_ref} -> 
+        {:ok, body} = :hackney.body(client_ref)
```
Comment (achempion) [inline lib/waffle/file.ex:137]: we should extract this into separate function

File: lib/waffle/file.ex [hunk 2]
```diff
@@ -24,6 +24,7 @@ defmodule Waffle.File do
     filename = uri.path |> Path.basename() |> URI.decode()
 
     case save_file(uri, filename) do
+      {:ok, filename, local_path} -> %Waffle.File{path: local_path, file_name: filename, is_tempfile?: true}
```
Comment (achempion) [inline lib/waffle/file.ex:27]: I would like to keep the order, i.e. `{:ok, local_path, filename}`.

**Comment by ndarilek** [comment]
Thanks for the feedback! I've addressed everything but the testcase 
locally, which I'll get to on Monday.

**Comment by ndarilek** [comment]
I just pushed a failing test case. Could you please help me understand 
how to make this pass?


In particular, my thinking was to mock `Waffle.File.content_disposition` 
to return a filename, which in turn triggers the 3-argument tuple 
response. But I get:


```

      ** (ErlangError) Erlang error: {:undefined_function, {Waffle.File, 
:content_disposition, 1}}

```


Not sure if the issue is that the function is private, but ideally it 
should be.


I'm trying to get this in as a fix for a client app, so I can't spend 
much time researching this. If I'm missing some easy fix then please let 
me know and I'll update my previous WIP commit.


Thanks!

**Comment by achempion** [comment]
You can't really mock private functions in Elixir because they didn't
really exist.

They get compiled directly into calling function on the compilation
stage.

I think the reasonable way is to mock `:hackney_headers.new(headers)`
function in order to provide fake headers which will contain
content-disposition header with a completely new name and then check
that file with that name get saved.

**Comment by ndarilek** [comment]
Thanks for that clarification--I thought I was misusing the mock 
library. I'll give that a shot soon and let you know how it goes.

**Comment by ndarilek** [comment]
I tried replacing the mock with:

```
      {:hackney_headers,
        [],
        [new: fn ->
          IO.puts("Called")
          "favicon.ico"
        end]
      },
```

Just wanted to see if the mock was even called. No dice. Tried with a single-argument function too (I.e. `fn _ ->`).

Help very much appreciated. Our fork with these changes definitely works. I appreciate that you'd like a testcase for this--I just don't see any that operate this deep in the HTTP stack.

Thanks.

**Comment by ndarilek** [comment]
I'm happy to provide the testcase, but need help figuring out how to mock this method. I need to get this into production, so I'll either need to figure out how to merge this, or withdraw it and fix it in our field-specific uploader implementation so I'm once again on an actual release.

I also need to investigate and fix an issue where URL fields don't follow redirects, so there too I'll need guidance on how to structure the testcase. I've spent more time on this and can't seem to generate headers to make the test pass.

Thanks for your time.

**Comment by achempion** [comment]
I'll finish the testcase soon and I'll merge your work to the master branch. 

Thank you for your help and participation.

**Comment by ndarilek** [comment]
Thanks! Assuming it can be generalized, I'll use your testcase to inform 
future mocking efforts. I need to work on following redirects next.

**Comment by achempion** [comment]
It looks like you closed the issue and removed all the proposed changes.

**Comment by ndarilek** [comment]
Crap, I thought I'd done a rebase/force-push. Let me grab this repo out of my backups and push it again. Sorry.

**Comment by ndarilek** [comment]
I've repushed the changes to my [master branch](https://github.com/ndarilek/waffle). Again, apologies for the mess. I don't think I can reopen this PR, but the changes should be present.

**Pull request closed without merging**